[홍동욱] TodoList with Ajax

doonguk opened the pull request:
## 구현사항

- [x] todoItem 불러오기
- [x] todoItem 추가하기
- [x] todoItem 삭제하기
- [x] todoItem complete하기
- [x] todoItem contents 내용 수정하기
- [x] user별 투두리스트 불러오기
- [x] 데이터를 불러오기전 로딩바를 이용해, 사용자가 데이터가 불러와지고 있다는 것을 보여줍니다.
- [x] fetch api 사용하는 부분을 `async` `await`을 사용하여 리팩토링합니다.
- [x] github issue에서 라벨을 붙이는 것처럼, 우선순위에 따라서 badge를 추가합니다.
- [x] ES6 impot & export를 이용해 자바스크립트 파일을 리팩토링한다.

라벨 요구사항은 이해를 잘 하지 못 하여 조금 더 고민해보겠습니다 : )

## 질문사항

유저 이름을 클릭했을 때 클릭한 유저 버튼을 `active`하는 로직에 대해서 고민을 하였습니다. 제가 구현한 방법은 유저 이름을 클릭했을 때 `currentUser`라는 변수에 유저 이름을 업데이트 시키고 유저 컴포넌트를 다시 렌더링시켜 `currentUser`와 user 데이터의 이름이 같으면 `acitve` 클래스를 추가하는 방식입니다. 근데 이 방법은 버튼을 클릭할 때 계속 유저 컴포넌트를 처음부터 다시 그려서 비효율적인 것 같디고 생각합니다😥 좋은 아이디어 있으시면 공유 부탁드립니다 : )

wmakerjun committed:
refactor(typo): fix typo

doonguk committed:
1주차 코드 가져오기

doonguk committed:
fetchManager add(fetch 함수 모듈화)
TodoFilter Component delete(2주차 과제와 관계없다고 판단)
User Component add

doonguk committed:
유저 버튼 클릭 > 해당Todo 가져오기

doonguk committed:
Toggle, Delete API 연동

doonguk committed:
Edit API 연동

doonguk committed:
1. 불필요한 코드 지우기.
2. class 이름, 태그 이름 상수화

doonguk committed:
로딩 컴포넌트

wmakerjun committed:
feat(priority): add chip selection to show priority

doonguk committed:
1. TodoFilter 컴포넌트 다시 추가. hash를 이용하여 필터링 구현

doonguk committed:
전체삭제 API 연동

doonguk committed:
1. TodoItem 전체 삭제 구현
2. Priority API 연동, 기능 구현

doonguk commented:
1. 투두 리스트를 필터링하는 컴포넌트를 다시 추가하였습니다. 필터링할 때 로직은 [호준님의 미션1 PR 소영님의 커멘트](https://github.com/next-step/js-todo-list/pull/1)를 참고하여 hash를 사용해 구현하였습니다😀
2. 추가된 API인 전체삭제와 우선순위 설정 기능을 구현하였습니다.

eastjun-dev commented on App.js:
```diff
@@ -0,0 +1,206 @@
+import { httpMethod, filterStatus, className } from './utils/constants.js'
+import fetchManager from "./api/api.js"
+
+const delay = (ms) => new Promise((resolve) => setTimeout(() => resolve(), ms))
```
👍

eastjun-dev commented on App.js:
```diff
@@ -0,0 +1,206 @@
+const getTodoHash = (todos) => {
+  return {
+    [filterStatus.ALL]: todos,
+    [filterStatus.ACTIVE]: todos.filter(({isCompleted}) => isCompleted === false),
+    [filterStatus.COMPLETED]: todos.filter(({isCompleted}) => isCompleted === true)
```
단순 `true`, `false`를 비교할 경우  아래와 같이 표현하는 경우가 많더라구요 😄 취향이나 팀마다의 차이이니 이런 방식도 있구나 하고 알아주시면 좋을 것 같아요 😉 
```javascript-light
todos.filter(({isCompleted}) => !isCompleted),
todos.filter(({isCompleted}) => !!isCompleted)
```

doonguk replied:
넵! 알려주셔서 감사합니당😄 상황에 맞게(팀에 맞게) 적절하게 사용할게요!

kuongee commented on App.js:
```diff
@@ -0,0 +1,206 @@
+
+const getTodoHash = (todos) => {
+  return {
+    [filterStatus.ALL]: todos,
```
저도 덕분에 배워갑니다:)

kuongee commented on api/api.js:
```diff
@@ -0,0 +1,25 @@
+import { API_URL } from "../utils/constants.js"
+
+export default async function fetchManager(args){
```
fetchManager라고 공통으로 만들고 일관되게 API를 호출하는 점이 인상적이었습니다!
나중에 저도 리팩토링할 때 공통 모듈로 만들어보아야겠어요:)
그리고 한가지 제 생각을 덧붙이자면, 아무래도 fetch라는 함수를 썼기 때문에 fetchManager라고 이름이 정해진 거 같다고 생각하는데 맞나요?
아무래도 fetchManager만 본다면 무언가 가져온다는 느낌이 강해서요.
사실 삭제도 하고 수정도 할 수 있는데 말이에요.
그래서 requestManager? 같은 네이밍은 어떨지 의견 남겨봅니당~

doonguk replied:
네 맞아요!! `fetch`와 관련된 기능을 관리하는 함수명을 생각하다가 fetchManager 라고 했었어요😂 생각해보니 지수님 의견 처럼 '가져오다'라는 뜻이 더 강한 것 같아요! requestManager로 수정하겠습니다 : )

kuongee commented on api/api.js:
```diff
@@ -0,0 +1,25 @@
+  if (headers) {
+    options.headers = { ...options.headers, ...headers }
+  }
+  console.log('[REQUEST]') // 로그 확인용
```
확인용 로그가 구체적이어서 확인할 때 너무 좋을 거 같아요👍
근데 아무래도 디버깅용으로 사용될테니 디버깅 때만 사용될 수 있으면 좋겠다는 생각이 드는데요,
사실 저도 디버깅 때만 콘솔을 남길 수 있는 방법은 알지 못해 코멘트를 남기지 말까하다가
혹시 다른 분들이 아신다면 답변을 남겨주시길 기대하는 마음으로 남깁니다!

잠깐 구글링해보니까 DEBUG라는 글로벌 변수를 두어서 로그를 남길지 말지 결정한다는데 더 좋은 방법이 있을까요?

doonguk replied:
NodeJS 위에서 빌드를 하게 된다면 Node의 환경변수(`process.env`)를 이용한 방법이 생각이 나지만, 브라우저에서 바로 실행되는 JS로는 저도 구글링을 해봐야 할 것 같아요!! 아니면 내일 정기 미팅 때 한번 스터디원 분들에게 여쭤보기 좋은 질문 같아요 : )
좋은 지적  감사합니다👍

kuongee commented on components/Loading.js:
```diff
@@ -0,0 +1,23 @@
+import {checkSelector} from "../utils/validations.js"
```
{ checkSelector } 괄호 사이에 빈칸을 다른 파일들과 일관성있게 만들어주시면 좋을 거 같아요:)
다른 몇몇 파일들에도 있던데 확인 부탁드립니당!
저는 VScode를 사용하고 있는데 formatOnSave 기능으로 이런 띄어쓰기를 자동으로 맞춰주고 있거든요.
뭔가 이것도 설정했었던 거 같은데 지금 당장 못찾겠네요ㅠㅠ
https://vscode.tistory.com/entry/Prettier-Code-formatter
이 블로그에서 bracketSpacing  -> 이거인 거 같기도 하구.... 혹시 찾게 되면 다시 코멘트 남길게요!

doonguk replied:
고맙습니다! 지수님이 말씀해주신 `bracketSpacing` 옵션이 맞아요 : )
저는 VsCode 대신 다른 IDE를 쓰고 있어서 프리티어를 이용해서 일관성있게 수정하였습니다~!😆
친절하게 링크까지 달아주셔서 감사해요👍👍👍

kuongee commented on components/TodoList.js:
```diff
@@ -0,0 +1,133 @@
+        e.target.tagName === tagName.INPUT &&
+        e.target.className === className.EDIT
+      ) {
+        e.target.selectionStart = e.target.value.length
```
이게 혹시 포커스 될 때 문장/단어 가장 끝을 가리키도록 하는 그건가요?

doonguk replied:
네 그렇습니다!

kuongee commented on components/index.js:
```diff
@@ -0,0 +1,8 @@
+export { default as Header } from './Header.js'
```
너무 좋은 방법이네요!!!!! 배워갑니다~ 저도 리팩토링 때 적용해보겠습니당😀

kuongee replied:
이대로 코드를 정리하고 App.js의 1번째 줄에 있는 것처럼 수정하고
` npx http-server ` 로 로컬 서버 띄우고 실행시켜보니까 
```
Failed to load module script: The server responded with a non-JavaScript MIME type of "text/html". 
Strict MIME type checking is enforced for module scripts per HTML spec.
```
이런 에러가 나서요ㅠㅠ 근데 동욱님 코드 실행시켜보니까 같은 에러가 나네요......ㅜㅜ
어떻게 실행시키고 계신가요??😥

kuongee replied:
아래 동준님의 리뷰를 보고 힌트를 얻어 './components/index.js' 로 바꿔서 실행했습니다:)

doonguk replied:
저는 로컬서버를 띄우지 않고 index.html 파일을 직접실행 하여 브라우저에 띄우는 방식으로 했더니 위와 같은 오류가 뜨지 않았네요..몰랐습니다ㅜㅜ `./components/index.js`로 수정할게요!!😂

kuongee reviewed (commented):
동욱님 코드 보면서 '오 이렇게도 짤 수 있구나'라고 생각하면서 보았어요.
그 와중에 의견도 남기고 질문도 남겼는데 참고해주시면 감사하겠습니다.
혹시 제가 잘못 이해하여 남긴 코멘트가 있을 수도 있는데 이해해주시면 감사하겠습니당:)

제꺼 다 구현한 다음에 리뷰하려고 했는데 그럼 너무 늦어질거 같아서😁
구현하면서 느낀점이 있을테니 다 하고 나서 또 다시 한번 보러오겠습니다~

doonguk committed:
1. 프리티어 적용 > 일관성 있는 Code Convention 적용
2. fetchManager > requestManager로 수정
3. 확인용 로그를 개발환경에서만 확인하는 것은 조금 더 구글링 해보겠습니다 : )

eastjun-dev commented on App.js:
```diff
@@ -0,0 +1,206 @@
+    })
+
+    this.$removeAllBtn = document.querySelector(`.${className.REMOVE_ALL}`)
+    this.$removeAllBtn.addEventListener('click', () => {
```
아래와 같이 작성해도 괜찮을 것 같아요 😄 
```javascript
 this.$removeAllBtn.addEventListener('click', this.onDeleteAll)
```

eastjun-dev commented on App.js:
```diff
@@ -0,0 +1,206 @@
+    this.$todoList.setState(this.username, renderTodos)
+    this.$todoCount.setState(
+      renderTodos.length,
+      renderTodos.filter(({isCompleted}) => isCompleted === true).length
```
아래와 같이 작성해도 괜찮을 것 같아요 😄
```javascript
renderTodos.filter(({isCompleted}) => isCompleted).length
```

eastjun-dev commented on App.js:
```diff
@@ -0,0 +1,206 @@
+import { Header, User, TodoInput, TodoList, TodoCount, TodoFilter, Loading } from './components'
```
로컬에서 실행하려고 하면 `components`가 불려오지 않는데 혹시 확인해보시면 좋을 것 같아요 👀

eastjun-dev replied:
저는 로컬에서 아래와 같이 작성해야 정상적으로 동작하네요 ㅎㅎ
```javascript
import { Header, User, TodoInput, TodoList, TodoCount, TodoFilter, Loading } from "./components/index.js";
```

doonguk replied:
로컬서버를 띄우는 방식과 index.html 파일을 브라우저에서 실행시키는 방식에서 차이가 있었나 봐요 : ) 저는 후자의 방식으로 개발을 해서 오류가 나오지 않았었나 보네요ㅜㅜ 동준님이 알려주시지 않았다면 지수님이 리뷰하실 때 어려움이 있을 뻔했습니다.`components/index.js`로 수정하겠습니당 감사해요😆

eastjun-dev commented on App.js:
```diff
@@ -0,0 +1,206 @@
+  this.onToggle = async (itemId) => {
+   try {
+     await requestManager({
+       method: httpMethod.PUT,
```
method를 상수로 재활용하는 방식으로 하시다니 너무너무 좋은 것 같아요 👍 
여기에 보통 제 경우에는 실제 로직 내부 안에서 api를 직접 다 적기 보다는, 한 곳에서 관리하는게 다른 컴포넌트에서 동일한 api를 호출하더라도 재사용할 수가 있어서 더 좋은 것 같더라구요 😄 
그래서 보통 아래와 따로 분리해서 사용하는데 동욱님이 생각하기에 괜찮다고 생각하시면 한 곳에서 관리할 수있는 방법으로 분리해도 좋을 것 같아요 😄 

```javascript
const METHOD = {
  PUT(data) {
    return {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        ...data
      })
    }
  },
  POST(data) {
    return {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        ...data
      })
    }
  },
  DELETE() {
    return {
      method: 'DELETE'
    }
  }
}

const api = (() => {
  const request = (uri, config) => fetch(uri, config)

  const requestWithReturn = (uri, config) => request(uri, config).then((data) => data.json())

  const station = {
    get(id) {
      return requestWithReturn(`/stations/${id}`)
    },
    getAll() {
      return requestWithReturn(`/stations`)
    },
    create(data) {
      return requestWithReturn(`/stations`, METHOD.POST(data))
    },
    update(data, id) {
      return request(`/stations/${id}`, METHOD.PUT(data))
    },
    delete(id) {
      return request(`/stations/${id}`, METHOD.DELETE())
    }
  }
  
  return {
    station
  }
})()

export default api
```

doonguk replied:
정말..엄청난 방법입니다😱
저도 어떻게든 재사용하려고 노력했었는데, 이렇게 API 자체를 재사용하는 방법이 훨씬 더 좋은 방법인 것 같아요! 감사합니다 동준님😆😆

eastjun-dev commented on components/Loading.js:
```diff
@@ -0,0 +1,23 @@
+
+  this.render = () => {
+    this.$target = document.querySelector(selector)
+    this.$target.innerHTML = `
```
template을 해당 컴포넌트 안에서 작성하는 것도 좋지만, `templates.js`와 같은 파일을 만들어 한 곳에서 관리하면 어떨까요~? 그러면 마크업 구조가 비슷한 컴포넌트를 새로 만들 때 기존에 작성한 template들을 재활용하기에도 좋을 것 같아요 😄

eastjun-dev commented on components/TodoInput.js:
```diff
@@ -0,0 +1,28 @@
+  }
+
+  this.bindEvent = () => {
+    const keyPressHandler = (e) => {
```
`keyPressHandler`라는 네이밍이 다소 추상적일 수 있을 것 같아 보여요 😭 
혹시 `onAddTodoItemHandler`와 같이 어떤 역할을 할지 보다 명확한 네이밍을 같이 고민해보면 어떨까요? 🤔

doonguk replied:
넵! 이벤트 핸들러 함수의 네이밍을 할 때에는 기능에 관해서 고민하지 않았는데, 다른 함수들 처럼 기능과 관련된 네이밍을 하는게 좋을 것 같아요. 피드백 감사합니다 : )

eastjun-dev commented on components/TodoList.js:
```diff
@@ -0,0 +1,133 @@
+    this.$target.addEventListener('click', clickEventHandler)
+    this.$target.addEventListener('dblclick', dblclickEventHandler)
+    this.$target.addEventListener('keyup', keyUpEventHandler)
+    this.$target.addEventListener('focusin', focusInEventHandler) // 맨 마지막 글자에 focus
```
👍

eastjun-dev commented on components/TodoList.js:
```diff
@@ -0,0 +1,133 @@
+    const keyUpEventHandler = (e) => {
+      if (e.key === keyName.ESC) {
+        const li = e.target.closest('li')
+        li.classList.remove(className.EDITING)
```
45~46 line과 반복되는 것 같은데 중복을 줄이면 더 좋을 것 같아요 😄

eastjun-dev commented on components/TodoList.js:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  this.setState = (username, todos) => {
+    this.username = username
```
혹시 이 `this.username`은 어디서 사용되나요? 🤔

doonguk replied:
뜨끔.. 불필요한 변수였네요. 수정하겠습니다!🤣

eastjun-dev commented on components/TodoFilter.js:
```diff
@@ -0,0 +1,24 @@
+
+  this.bindEvent = () => {
+    this.$target.addEventListener('click', (e) => {
+      if (e.target.tagName.toLowerCase() === 'a') {
```
`tagName` 비교하는 로직에서 다른 부분들은 `tagName.INPUT`과 같이 상수를 재사용하는 방식으로 이용하셨는데, 여기서는 직접 `a`라고 작성해주신 이유가 있을까요? 🤔

doonguk replied:
처음에 러프하게 작성한 뒤, 후에 상수로 따로 빼서 작성하다가 빼먹었네요! 알려주셔서 감사합니다 : )

eastjun-dev commented on utils/storage.js:
```diff
@@ -0,0 +1,12 @@
+export const storage = {
```
혹시 이 storage 객체를 사용하는 곳이 있을까요~? 🤔

doonguk replied:
이것도 불필요한 객체입니다. 수정하겠습니다🤣

eastjun-dev commented on components/TodoList.js:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    const changeEventHandler = (e) => {
+      if (e.target.tagName === tagName.SELECT) {
```
저 같은 경우는 if문이 Nested하게 들어가는 경우가 생기면 나중에 로직이 점점 복잡해지고, 방어코드 짜면서 헷갈리는 경우가 많이 생기더라구요 😭 
그래서 보통 아래처럼 작성하곤 해요!
```javascript
if (e.target.tagName !== tagName.SELECT) {
  return
}
const li = e.target.closest('li')
const id = li.dataset.id
if (e.target.value !== 0) {
  onSetPriority(id, e.target.value)
}
...
```

doonguk replied:
꿀팁 감사합니다 😆 방금 피드백 주신대로 수정을 하였는데 코드도 훨씬 깔끔해진 것 같아요!!

eastjun-dev commented on components/User.js:
```diff
@@ -0,0 +1,39 @@
+  this.setState = () => {
+    this.$target.innerHTML = this.users.map(({ _id, name }) => {
+      return `<button class="ripple ${this.currentUser === name ? 'active' : ''}">${name}</button>`
+    }).join()
```
`join("")`이렇게 해줘야 아래 이미지처럼 버튼 사이에 있는 콤마가 사라질 것 같아요 😄 
<img width="498" alt="스크린샷 2020-06-29 오전 9 23 22" src="https://user-images.githubusercontent.com/56821976/85967792-3f1e3f00-b9fe-11ea-9330-9404aa640083.png">

doonguk replied:
오류 지적해주셔서 감사합니다! 수정했습니당☺️

eastjun-dev reviewed (changes requested):
## 동욱님~~~~ 😄

component의 상세 분리나, api의 사용, 한 메서드당 길지 않은 로직을 통해 읽기 좋은 코드까지!!!
코드리뷰하면서 저도 많이 공부가 되었어요 😄참여 감사합니다.
보다보니 2가지 정도 조금 다른 생각이 있어서 피드백으로 남겨보려고 해요 :)

### 1. App.js의 역할 나누기
App.js의 경우에는 컴포넌트를 통틀어 전체 상태가 관리되는 컴포넌트일텐데요. 전체 상태를 관리하는 로직 외에도 api 통신이 구현되어있던데 각 api 통신은 해당 컴포넌트안에서 구현되는게 맞는게 아닐까라는 생각이 들어요~

예를들면 `App.js`에 이렇게 구현되어 있고

```javascript
const todoInput = new TodoInput({ 
	$target: document.querySelector('#todo-input'), 
	onAdd: (text) => {
	  const nextData = [...this.data]
	  nextData.push({
	    text,
	  })    
	  this.setState(nextData)
	}
})
```

`TodoInput.js`에 아래와 같이 구현하는 것이죠!


```javascript
$target.addEventListener('keydown', (e) => {
    try {
      await fetchManager({
        method: httpMethod.POST,
        path: `/api/u/${this.username}/item`,
        body: { contents: text }
      });
      onAdd()
    } catch (e) {
      console.error(e);
    }
	})
})
```

### 2. template 분리
각 컴포넌트에 template을 작성하는 것도 좋지만, 마크업을 재사용하기에는 어려운 구조로 보여서 저는 보통 templates.js에 따로 분리해서 import해서 사용하는데 어떻게 생각하실까요? 🤔 

2주차 미션 진행하느라 너무 고생하셨구요!!! 앞으로 🌟 더 빛나는🌟 성장 같이해요!

pocojang commented on utils/validations.js:
```diff
@@ -0,0 +1,4 @@
+export const checkSelector = (selector) => {
```
아마 셀렉터가 잡지 못하는 요소는 브라우저에서 알아서 에러를 던져줄텐데
커스텀 에러가 필요한게 아니라면 무의미할 수 있습니다!

doonguk replied:
@devJang  앗..생각해보니 그러네요 delete 하겠습니다😭 장님 질문이 있습니다. 보통 커스텀 에러는 어떤 경우에 사용하는 건가요?

pocojang replied:
에러 코드나 메세지를 공통화해서 사용하거나 일관적인 처리를 위해 따로 만드시더라구요 ㅎㅎ
직접 만든 오류를 분류할 수 있으니까요!

https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Error

doonguk replied:
아아 그렇군요🙂 고맙습니다 보내주신 링크도 읽어볼게요~!!

pocojang commented on components/TodoList.js:
```diff
@@ -0,0 +1,133 @@
+        e.target.className === className.TOGGLE
+      ) {
+        onToggle(id)
+      } else if (e.target.tagName === tagName.BUTTON) {
```
`else if` => `if` 변경 가능해보이며
함수의 매개변수에서 바로 ({ target })으로 꺼내 사용한다면 `e.target` => `target` 사용가능해보이네요!

doonguk replied:
`else if` > `if`, `e.target` > `target`으로 수정하였습니다😆
@devJang 장님 혹시 `if`로 코드를 변경함으로써 얻는 이점에 대해서 여쭤볼 수 있을까요?😮😮

pocojang commented on utils/constants.js:
```diff
@@ -0,0 +1,36 @@
+
+export const STORAGE_KEY = 'TODO' //for localstorage
+
+export const httpMethod = {
```
상수라면 객체 명도 `SNAKE_CASE`를 써서 동일하게 케이스를 가져가는 게 낫지 않을까요?!

doonguk replied:
일관성있게 작성하는게 맞다고 생각해요! `SNAKE_CASE`로 수정하였습니다 장님 피드백 감사해요 😊

doonguk committed:
1. filter 조건문 간소화
2. 불필요한 파일, 함수, 변수 삭제
3. 오타 및 오류 수정

doonguk committed:
1. App 컴포넌트 기능 분리 > API 요청에 관한 비지니스 로직을 하위 컴포넌트가 담당하게 구현하였습니다.
2. TodoList 컴포넌트 focusin 이벤트 삭제 > 지수님 코드를 참고해서 focus 와 텍스트 끝에 자동으로 포커스가 되게 구현하였습니다.
3. 투두 아이템을 수정할 때 label 태그의 text 값을 가져올 때 `e.target.innerText`를 사용하면 select 노드의 text 값 까지 가져오는 오류가 있어서 `e.target.childNodes[2].textContent`를 사용하여 text 노드의 값을 구했습니다.

doonguk committed:
Template.js > 컴포넌트의 HTML 템플릿 분리

doonguk committed:
devJang님 피드백 반영

kuongee reviewed (approved):
이번 미션 수행도 너무 수고 많으셨어요! 👍
피드백 반영도 꼼꼼하게 해주셔서 감사합니다:)

doonguk committed:
1. 동준님 피드백 > API를 모듈화하여 재사용성을 높임
2. 이벤트 핸들러 함수들 네이밍 수정 (좀 더 명확하게)
3. TodoItem 수정 이벤트 핸들러 > ESC 와 ENTER 모두 아니면 fast fail로 변경
